### PR TITLE
Add word boosting to NemotronStreamingASR

### PR DESCRIPTION
**Pull request by soloish90**
Submitting this as a draft as i was hesitant to even submit it, as this code isn't my area of expertise and not sure what the policy on this repo is. Hopefullly useful. Completed with a large dose of help from Codex. That said it IS working well in my current alpha of my mac realtime transcription app Babble(https://github.com/soloish90/babble). That app also uses the Nemotron ASR from speech-swift.

------------------------
This adds word boosting to Nemotron streaming sessions: you give the session a list of phrases (names, technical terms, product words), and the decoder gets biased toward them while it decodes. The approach is shallow fusion over the joint logits. Boost phrases are tokenized and built into a small prefix trie with failure links, and at each decode step the configured boost is added to the scores of tokens that continue a known phrase, scaled up the deeper into the phrase you get.

```
let suggestions = model.wordBoostingSuggestions(for: ["Voxtral", "Qwen"])
let config = WordBoostingConfig(phrases: suggestions.map { .init($0.phrase, boost: $0.suggestedBoost) })
let session = try model.createSession(language: "en-US", wordBoosting: config)
```
One detail that turned out to matter a lot: boosting only works if the trie is built from the same token sequences the decoder actually emits, which means tokenizing phrases with the real SentencePiece model rather than greedily segmenting against vocab.json. I tested both on a sample of out-of-vocabulary words and the greedy segmentation disagreed with SentencePiece on 10 of 14 — for example "voxtral" is ▁ vo x tra l under the real tokenizer but ▁v o x tra l greedily. They diverge on the first token, so a trie built from the greedy path just never fires. The loader picks up tokenizer.model from the bundle directory when present and falls back to vocab.json segmentation when it isn't (with a status you can inspect via wordBoostingTokenizerStatus), but the fallback should be considered degraded — the file is small and worth shipping.

TLDR: i added tokenizer.model extracted from NVIDIA's .nemo archive, to make this work.

There's a wordBoostingSuggestions(for:) helper that recommends a boost strength per phrase based on how fragmented its tokenization is. The thresholds are still being tuned and tested — high boosts can get pushy and greedily override other words (vase becomes voxtral), and I'm still dialing in where the lines should sit from day-to-day use.

Tests are in WordBoostingTests (tree scoring, fallback links, selection, tokenizer), and there's a section in docs/inference/nemotron-asr-streaming.md covering usage and boost ranges. 

All of this is initial and experimental. just thought it might be a good idea to post a draft PR and get early feedback...  if it is useful at all, etc...

  ---

### CODE REVIEW

**Comment by ivan-digital** [comment]
Took a pass through the review punch list (since maintainer-edit was on) — pushed `33d564a` directly to `word-boosting`. Nothing controversial, all localized:

**1. Reuse `AudioCommon.SentencePieceModel`.** The 115-line `ProtobufReader` in `SentencePieceUnigramTokenizer.swift` is gone — it now parses through the shared reader that OmnilingualASR / PersonaPlex / SpeechWakeWord already use. Two parsers were going to drift.

**2. Filter Viterbi by piece type, not by `id != 0`.** The previous gate assumed `<unk>` was at index 0 (true for Nemotron 3.5 but not portable, and it let `control` / `byte` / `unused` pieces still be selected as spans). Now `guard !pieces[id].isControlOrUnknown` does the job uniformly.

**3. Back-pointers instead of prefix copies in Viterbi.** Was `O(N²)` accepted spans × `O(N)` `Array<Int>` copy per acceptance = `O(N³)` ints in flight. Now `(start, id)` back-pointer + reverse walk at the end. Also added a 256-scalar phrase cap so a clipboard paste can't pin the calling thread.

**4. Suggestion heuristic.** `WordBoosting.swift:158, 169` — gated `hasShortShape` on `bestTokenCount > 1`. The previous rule routed `AI` / `iOS` / `kHz` to `.hard / 1.25` because they're ≤4 scalars, even though they tokenize to a single in-vocab piece. Added a unit test asserting `.easy / 0.75` for the single-piece short case.

**5. Cleanup.** Dropped `WordBoostingTree.Node.tokens` (full prefix stored at every node, never read externally) and the unused `tokens(at:)` accessor.

**6. Fallback log.** `NemotronStreamingASR` now emits an `AudioLog.modelLoading.warning` when `tokenizer.model` is missing and word boosting falls back to greedy vocab segmentation — your 10/14 OOV observation made it clear this shouldn't be silent.

49 `NemotronStreamingASR*` tests pass / 0 failures / 3 skipped (the cached-bundle parity tests skip without the downloaded `tokenizer.model`).

Going to push another update with the open follow-ups — particularly an E2E test that decodes real audio with and without boosting and asserts an OOV term actually flips, plus a `scripts/` parity check against Python `sentencepiece` over a corpus to catch normalization drift if the vocabulary ever changes.

**Comment by soloish90** [comment]
nice. just catching up with this. Thanks for finishing/merging.
